File: src/main/java/ludwig/ide/PackageTreeView.java
package ludwig.ide;

import javafx.scene.control.*;
import ludwig.changes.*;
import ludwig.model.*;
import ludwig.workspace.Workspace;

import static java.util.Collections.singletonList;
import static ludwig.utils.NodeUtils.isReadonly;

class PackageTreeView extends TreeView<NamedNode> {
    private final Workspace workspace;

    PackageTreeView(Workspace workspace) {
        super(createRoot(workspace));
        this.workspace = workspace;
        setShowRoot(false);

        setContextMenu(ContextMenuFactory.menu(new Actions()));

        setPrefWidth(120);
    }

    private static TreeItem<NamedNode> createRoot(Workspace workspace) {
        TreeItem<NamedNode> root = new TreeItem<>();

        for (ProjectNode projectNode : workspace.getProjects()) {
            TreeItem<NamedNode> projectItem = new TreeItem<>(projectNode);

            root.getChildren().add(projectItem);

            for (Node packageNode : projectNode.children()) {
                processPackage(projectItem, (PackageNode) packageNode);
            }
        }

        return root;
    }

    private static void processPackage(TreeItem<NamedNode> parent, PackageNode packageNode) {
        TreeItem<NamedNode> packageItem = new TreeItem<>(packageNode);
        parent.getChildren().add(packageItem);

        for (Node node : packageNode.children()) {
            if (node instanceof PackageNode) {
                processPackage(packageItem, (PackageNode) node);
            }
        }
    }


    void select(NamedNode packageNode) {
        getSelectionModel().select(find(packageNode));
    }

    private TreeItem<NamedNode> find(NamedNode node) {
        if (node == null) {
            return null;
        }
        TreeItem<NamedNode> parentItem = node.parent() == null ? getRoot() : find((NamedNode) node.parent());
        return parentItem.getChildren()
            .stream()
            .filter(i -> i.getValue() == node)
            .findFirst()
            .orElse(null);
    }

    void refresh() {
        setRoot(createRoot(workspace));
    }

    public class Actions {
        public void newPackage() {
            TreeItem<NamedNode> selectedItem = getSelectionModel().getSelectedItem();
            if (selectedItem != null) {
                Node parent = selectedItem.getValue();

                if (isReadonly(parent)) {
                    return;
                }

                TextInputDialog dialog = new TextInputDialog();
                dialog.setTitle("Add a package");
                dialog.setHeaderText("");
                dialog.setContentText("Package name");

                dialog.showAndWait().ifPresent(name -> {
                    PackageNode packageNode = new PackageNode().name(name).id(Change.newId());
                    InsertNode insert = new InsertNode()
                        .node(packageNode)
                        .parent(parent.id());
                    workspace.apply(singletonList(insert));
                    select(packageNode);
                });
            }
        }

        public void deletePackage() {
            TreeItem<NamedNode> selectedItem = getSelectionModel().getSelectedItem();
            if (selectedItem != null) {
                Node packageNode = selectedItem.getValue();
                if (isReadonly(packageNode)) {
                    return;
                }
                NamedNode parent = (NamedNode) packageNode.parent();
                workspace.apply(singletonList(new Delete().id(packageNode.id())));
                select(parent);
            }
        }
    }
}

File: src/main/kotlin/ludwig/ide/PackageTreeView.kt
package ludwig.ide

import javafx.scene.control.TextInputDialog
import javafx.scene.control.TreeItem
import javafx.scene.control.TreeView
import ludwig.changes.Create
import ludwig.changes.Delete
import ludwig.changes.Rename
import ludwig.model.NamedNode
import ludwig.model.PackageNode
import ludwig.utils.NodeUtils
import ludwig.workspace.Workspace

internal class PackageTreeView : TreeView<NamedNode> {
    private val workspace: Workspace

    constructor(workspace: Workspace) : super(createRoot(workspace)) {
        this.workspace = workspace
    }

    init {
        isShowRoot = false

        contextMenu = ContextMenuFactory.menu(Actions())

        prefWidth = 120.0
    }


    fun select(packageNode: NamedNode?) {
        selectionModel.select(find(packageNode))
    }

    private fun find(node: NamedNode?): TreeItem<NamedNode>? {
        if (node == null) {
            return null
        }
        val parentItem = if (node.parent == null) root else find(node.parent as NamedNode?)
        return parentItem!!.children
                .stream()
                .filter { i -> i.value === node }
                .findFirst()
                .orElse(null)
    }

    fun recreateTree() {
        root = createRoot(workspace)
    }

    inner class Actions {
        fun newPackage() {
            val selectedItem = selectionModel.selectedItem
            if (selectedItem != null) {
                val parent = selectedItem.value

                if (NodeUtils.isReadonly(parent)) {
                    return
                }

                val dialog = TextInputDialog()
                dialog.title = "Add a package"
                dialog.headerText = ""
                dialog.contentText = "Package name"

                dialog.showAndWait().ifPresent { name ->
                    val create = Create().apply { nodeType = PackageNode::class.simpleName!!; this.parent = parent.id }
                    val rename = Rename().apply { this.name = name; nodeId = create.changeId }
                    workspace.apply(create, rename)
                    //                   select(packageNode)
                }
            }
        }

        fun deletePackage() {
            val selectedItem = selectionModel.selectedItem
            if (selectedItem != null) {
                val packageNode = selectedItem.value
                if (NodeUtils.isReadonly(packageNode)) {
                    return
                }
                val parent = packageNode.parent as NamedNode?
                workspace.apply(Delete().apply { nodeId = packageNode.id })
                select(parent)
            }
        }
    }
}

private fun createRoot(workspace: Workspace): TreeItem<NamedNode> {
    val root = TreeItem<NamedNode>()

    for (projectNode in workspace.projects) {
        val projectItem = TreeItem<NamedNode>(projectNode)

        root.children.add(projectItem)

        for (packageNode in projectNode) {
            processPackage(projectItem, packageNode as PackageNode)
        }
    }

    return root
}

private fun processPackage(parent: TreeItem<NamedNode>, packageNode: PackageNode) {
    val packageItem = TreeItem<NamedNode>(packageNode)
    parent.children.add(packageItem)

    for (node in packageNode) {
        if (node is PackageNode) {
            processPackage(packageItem, node)
        }
    }
}

